### ISSUE TITLE
Adding useful fields to admin/customer/list for exporting to xlsx/xml

### ISSUE THREAD

**Issue by rolandh** [nopSolutions/nopCommerce#3078]
The admin/customers/list view is lacking lots of useful fields. Eg it exports the countryId not the actual country string etc which is not useful for importing into other products.

I have made a new controller to override this function:
`        [HttpPost, ActionName("List")]
        [FormValueRequired("exportexcel-all")]
        [Route("/Admin/Customer/List")]
        public virtual IActionResult ExportExcelAll(CustomerListModel model)
        {`

I have then modified it to call my own version of ExportCustomersToXlsx()

This is not very efficient (eg calling GetCustomerByEmail 4x instead of once per customer etc) however it works great. It adds DOB, real country and state strings etc.

Can we have a new dropdown menu item (or a mechanism to add custom ones easily) that calls this code? The way I have implemented it currently so that I don't modify the original nop source is very hacky (eg overriding the url).

Any way we can get this code in the released version of nop? I use it for importing into my forum/facebook advert exclude list etc and it is very handy.

```
public virtual byte[] ExportCustomersToXlsx(IList<Customer> customers)
{
	//property array
	var properties = new[]
	{
		new PropertyByName<Customer>("CustomerId", p => p.Id),
		new PropertyByName<Customer>("CustomerGuid", p => p.CustomerGuid),
		new PropertyByName<Customer>("Email", p => p.Email),
		new PropertyByName<Customer>("Username", p => p.Username),
		new PropertyByName<Customer>("Password", p => _customerService.GetCurrentPassword(p.Id)?.Password),
		new PropertyByName<Customer>("PasswordFormatId", p => _customerService.GetCurrentPassword(p.Id)?.PasswordFormatId ?? 0),
		new PropertyByName<Customer>("PasswordSalt", p => _customerService.GetCurrentPassword(p.Id)?.PasswordSalt),
		new PropertyByName<Customer>("IsTaxExempt", p => p.IsTaxExempt),
		new PropertyByName<Customer>("AffiliateId", p => p.AffiliateId),
		new PropertyByName<Customer>("VendorId", p => p.VendorId),
		new PropertyByName<Customer>("Active", p => p.Active),
		new PropertyByName<Customer>("IsGuest", p => p.IsGuest()),
		new PropertyByName<Customer>("IsRegistered", p => p.IsRegistered()),
		new PropertyByName<Customer>("IsAdministrator", p => p.IsAdmin()),
		new PropertyByName<Customer>("IsForumModerator", p => p.IsForumModerator()),
		new PropertyByName<Customer>("CreatedOnUtc", p => p.CreatedOnUtc),
		//attributes
		new PropertyByName<Customer>("FirstName", p => p.GetAttribute<string>(SystemCustomerAttributeNames.FirstName)),
		new PropertyByName<Customer>("LastName", p => p.GetAttribute<string>(SystemCustomerAttributeNames.LastName)),
		new PropertyByName<Customer>("Gender", p => p.GetAttribute<string>(SystemCustomerAttributeNames.Gender)),

		new PropertyByName<Customer>("DateOfBirth", p => p.GetAttribute<string>(SystemCustomerAttributeNames.DateOfBirth)),

		new PropertyByName<Customer>("Company", p => p.GetAttribute<string>(SystemCustomerAttributeNames.Company)),
		new PropertyByName<Customer>("StreetAddress", p => p.GetAttribute<string>(SystemCustomerAttributeNames.StreetAddress)),
		new PropertyByName<Customer>("StreetAddress2", p => p.GetAttribute<string>(SystemCustomerAttributeNames.StreetAddress2)),
		new PropertyByName<Customer>("ZipPostalCode", p => p.GetAttribute<string>(SystemCustomerAttributeNames.ZipPostalCode)),
		new PropertyByName<Customer>("City", p => p.GetAttribute<string>(SystemCustomerAttributeNames.City)),
		new PropertyByName<Customer>("CountryId", p => p.GetAttribute<int>(SystemCustomerAttributeNames.CountryId)),
		new PropertyByName<Customer>("StateProvinceId", p => p.GetAttribute<int>(SystemCustomerAttributeNames.StateProvinceId)),

		new PropertyByName<Customer>("Country", p =>
		{
			var customer = _customerService.GetCustomerByEmail(p.Email);
			return customer?.Addresses?.FirstOrDefault()?.Country?.Name;
		}),
		new PropertyByName<Customer>("CountryAbbr", p =>
		{
			var customer = _customerService.GetCustomerByEmail(p.Email);
			return customer?.Addresses?.FirstOrDefault()?.Country?.ThreeLetterIsoCode;
		}),
		new PropertyByName<Customer>("State", p =>
		{
			var customer = _customerService.GetCustomerByEmail(p.Email);
			return customer?.Addresses?.FirstOrDefault()?.StateProvince?.Name;
		}),
		new PropertyByName<Customer>("StateAbbr", p =>
		{
			var customer = _customerService.GetCustomerByEmail(p.Email);
			return customer?.Addresses?.FirstOrDefault()?.StateProvince?.Abbreviation;
		}),

		new PropertyByName<Customer>("Phone", p => p.GetAttribute<string>(SystemCustomerAttributeNames.Phone)),
		new PropertyByName<Customer>("Fax", p => p.GetAttribute<string>(SystemCustomerAttributeNames.Fax)),
		new PropertyByName<Customer>("VatNumber", p => p.GetAttribute<string>(SystemCustomerAttributeNames.VatNumber)),
		new PropertyByName<Customer>("VatNumberStatusId", p => p.GetAttribute<int>(SystemCustomerAttributeNames.VatNumberStatusId)),
		new PropertyByName<Customer>("TimeZoneId", p => p.GetAttribute<string>(SystemCustomerAttributeNames.TimeZoneId)),
		new PropertyByName<Customer>("AvatarPictureId", p => p.GetAttribute<int>(SystemCustomerAttributeNames.AvatarPictureId)),
		new PropertyByName<Customer>("ForumPostCount", p => p.GetAttribute<int>(SystemCustomerAttributeNames.ForumPostCount)),
		new PropertyByName<Customer>("Signature", p => p.GetAttribute<string>(SystemCustomerAttributeNames.Signature))
	};

	return ExportToXlsx(properties, customers);
}

```

**Comment by skoshelev** [comment]
Hi @rolandh.  looked at your code, do I understand correctly that the main problem is that identifiers are exported instead of real names? Since there is only one CountryAbbr field in your list that is not in our list?

Also, I think the code:

```
new PropertyByName<Customer>("Country", p =>
{
	var customer = _customerService.GetCustomerByEmail(p.Email);
	return customer?.Addresses?.FirstOrDefault()?.Country?.Name;
})
```
can be written like this:

`new PropertyByName<Customer>("Country", p => p.Addresses?.FirstOrDefault()?.Country?.Name)`

since the parameter **p** already contains the user you need

**Comment by rolandh** [comment]
Yes so we currently have overridden the code to export everything we need. But I think it would be a useful feature to add to the main nop source where the actual country (and not the identifier) is exported to the XLSX sheet along with any other identifiers that have corresponding real descriptions.

**Comment by skoshelev** [comment]
Closed #3078
